src/lib/Microsoft.Fx.Portability.Azure/Storage/StorageBlob.cs

```diff
@@ -0,0 +1,147 @@
+        /// <returns></returns>
+        public async Task<bool> SaveToBlobAsync(string uniqueId, AnalyzeRequest request)
+        {
+            DateTime currentDate = DateTime.Now;
+            string containerName = GetCurrentMonthContainerName(currentDate);
+            CloudBlobContainer container = await GetBlobContainerAsync(containerName);
+
+            if (container == null)
+            {
+                return false;
+            }
+
+            byte[] content = request.SerializeAndCompress();
+            var containerPath = $"{currentDate.Day.ToString("00", CultureInfo.InvariantCulture)}/{uniqueId}";
+
+            CloudBlockBlob blob = container.GetBlockBlobReference(containerPath);
+            blob.Metadata.Add("Version", request.Version.ToString(CultureInfo.InvariantCulture));
+            await blob.UploadFromByteArrayAsync(content, 0, content.Length);
+            await blob.SetMetadataAsync();
+            return true;
+        }
+
+        public async Task<AnalyzeRequest> RetrieveFromBlobAsync(string uniqueId, string containerName)
+        {
+            CloudBlobContainer container = await GetBlobContainerAsync(containerName);
+            CloudBlockBlob blob = container.GetBlockBlobReference(uniqueId);
+
+            using (var blobStream = new MemoryStream())
+            {
+                await blob.DownloadToStreamAsync(blobStream);
+
+                var blobData = blobStream.ToArray();
+
+                return blobData.DecompressToObject<AnalyzeRequest>();
+            }
+        }
+
+        public async Task<IEnumerable<string>> RetrieveRequestIdsAsync()
+        {
+            var blobs = await GetBlobsAsync();
+            List<string> results = new List<string>();
+            foreach (var blob in blobs)
+            {
+                results.Add(blob.Name);
+            }
+
+            return results;
+        }
+
+        public async Task<AnalyzeRequest> RetrieveFromBlobAsync(string uniqueId)
+        {
+            var blobs = await GetBlobsAsync();
+            var blob = blobs.SingleOrDefault(b => b.Name == uniqueId);
+
+            if (blob == null)
+            {
+                return null;
+            }
+
+            using (var blobStream = await blob.OpenReadAsync())
+            {
+                return blobStream.DecompressToObject<AnalyzeRequest>();
+            }
+        }
+
+        private async Task<IEnumerable<CloudBlockBlob>> GetBlobsAsync()
+        {
+            var blobClient = _storageAccount.CreateCloudBlobClient();
+            var containers = await blobClient.ListContainersAsync(prefix: "container").ConfigureAwait(false);
+            var blobTasks = containers.Select(c => c.ListBlobsAsync());
+            var blobs = await Task.WhenAll(blobTasks).ConfigureAwait(false);
+            var query = blobs.SelectMany(b => b)
+                .OfType<CloudBlockBlob>();
+            return query;
+        }
+
+        public static string GetCurrentMonthContainerName()
```

Comment:
Do we use all these public methods? We should remove all the unused methods and keep only the ones in use to keep our new codebase clean

Reply:
I wasn't sure what will be needed in the future new services.  I will remove them now and we can always add them back.

---
In reply to: [178099148](https://github.com/Microsoft/dotnet-apiport/pull/598#discussion_r178099148) [](ancestors = 178099148)

Reply:
Fixed

---
In reply to: [178112261](https://github.com/Microsoft/dotnet-apiport/pull/598#discussion_r178112261) [](ancestors = 178112261,178099148)